File: Jasmine/JasmineCompiler/Src/DFIR_Compiler/DFIR_Compiler/IForLoop.cs
using System;
using System.Collections.Generic;


namespace DFIR_Compiler
{
    public class IForLoop:INode
    {
        private ILoopIndex iLoopIndex;
        private ILoopMax iLoopMax;
        private List<ITunnel> iTunnels;
        private List<INode> inputBorderNodes;
        private List<INode> outputBorderNodes;
        private List<ILeftShiftRegister> iLeftShiftRegisters;
        private List<IRightShiftRegister> iRightShiftRegisters;
        private IDiagram iDiagram;

        public IForLoop(int nodeId, int parentId, List<ITerminal> inputTerminals, List<ITerminal> outputTerminals, ILoopIndex _iLoopIndex, ILoopMax _iLoopMax, List<ITunnel> _iTunnels, List<ILeftShiftRegister> _iLeftShiftRegisters, List<IRightShiftRegister> _iRightShiftRegisters, IDiagram _iDiagram) : base(nodeId, parentId, inputTerminals, outputTerminals)
        {
            iLoopIndex = _iLoopIndex;
            iLoopMax = _iLoopMax;
            iTunnels = _iTunnels;
            inputBorderNodes = new List<INode>();
            outputBorderNodes = new List<INode>();
            iLeftShiftRegisters = _iLeftShiftRegisters;
            iRightShiftRegisters = _iRightShiftRegisters;
            iDiagram = _iDiagram;
            INodeType = "IForLoop";

            foreach (ITunnel i in iTunnels)
            {
                if (!i.GetIsInput())
                {
                    outputBorderNodes.Add(i);
                }
                else
                {
                    inputBorderNodes.Add(i);
                }
            }
            foreach (ILeftShiftRegister i in iLeftShiftRegisters)
            {
                inputBorderNodes.Add(i);
            }
            foreach (IRightShiftRegister i in iRightShiftRegisters)
            {
                outputBorderNodes.Add(i);
            }
            //The number of the input border nodes is the indegree of IForLoop
            indegree = inputBorderNodes.Count; 
            if (iLoopMax.GetIsInputConnected())
            {
                indegree++;
            }
        }

        public virtual IDiagram GetIDiagram()
        {
            return iDiagram;
        }
        public virtual ILoopIndex GetILoopIndex()
        {
            return iLoopIndex;
        }
        public virtual ILoopMax GetILoopMax()
        {
            return iLoopMax;
        }
        public virtual List<ITunnel> GetITunnels()
        {
            return iTunnels;
        }
        public virtual List<INode> GetOutputBorderNodes()
        {
            return outputBorderNodes;
        }
        public virtual List<INode> GetInputBorderNodes()
        {
            return inputBorderNodes;
        }
        public virtual List<ILeftShiftRegister> GetILeftShiftRegisters()
        {
            return iLeftShiftRegisters;
        }
        public virtual List<IRightShiftRegister> GetIRightShiftRegisters()
        {
            return iRightShiftRegisters;
        }


        public override void Print()
        {
            Console.WriteLine("------------------------------------");
            Console.WriteLine("IForLoop:");
            Console.WriteLine("NodeId: " + nodeId + ' ' + "ParentId: " + parentId);
            iLoopIndex.Print();
            iLoopMax.Print();
            for (int i = 0; i < iTunnels.Count; i++)
            {
                iTunnels[i].Print();
            }
            for (int i = 0; i < iLeftShiftRegisters.Count; i++)
            {
                iLeftShiftRegisters[i].Print();
            }
            for (int i = 0; i < iRightShiftRegisters.Count; i++)
            {
                iRightShiftRegisters[i].Print();
            }
            iDiagram.Print();
        }


    }
}

File: Wayne/WayneCompiler/src/Dfir_representation/IForLoop.java
package Dfir_representation;

import java.util.Vector;


public class IForLoop extends IStructure{
	   
	   private ILoopIndex iLoopIndex;
	   private ILoopMax iLoopMax;
	   private Vector<ITunnel> iTunnels;
	   private Vector<ILeftShiftRegister> iLeftShiftRegisters;
	   private Vector<IRightShiftRegister> iRightShiftRegisters;
	   private Vector<JNode> inputBorderNodes;
	   private Vector<JNode> outputBorderNodes;
	   private IDiagram iDiagram;
	   
	   public IForLoop(int nodeId, int parentId, Vector<ITerminal> inputTerminals,
			           Vector<ITerminal> outputTerminals,  ILoopIndex iLoopIndex, ILoopMax iLoopMax,
			           Vector<ITunnel> iTunnels,Vector<ILeftShiftRegister> iLeftShiftRegisters,
			           Vector<IRightShiftRegister> iRightShiftRegisters,IDiagram iDiagram){
		   super(nodeId,parentId,inputTerminals,outputTerminals);
		   JNodeType = "IForLoop";
		   this.iLoopIndex = iLoopIndex;
		   this.iLoopMax = iLoopMax;
		   this.iTunnels = iTunnels;
		   this.iLeftShiftRegisters = iLeftShiftRegisters;
		   this.iRightShiftRegisters = iRightShiftRegisters;
		   this.inputBorderNodes = new Vector<JNode>();
		   this.outputBorderNodes = new Vector<JNode>();
		   this.iDiagram = iDiagram;
		   
		   for(ITunnel i : this.iTunnels){
			   if(!i.getIsInput()){ 
				   outputBorderNodes.addElement(i);
			   }else{
				   inputBorderNodes.addElement(i);
			   }
		   }
		   for(ILeftShiftRegister i : this.iLeftShiftRegisters){
			   inputBorderNodes.addElement(i);
		   }
		   for(IRightShiftRegister i : this.iRightShiftRegisters){
			   outputBorderNodes.addElement(i);
		   }
		   this.indegree = this.inputBorderNodes.size();
		   if(this.iLoopMax.getIsInputConnected()){
			   this.indegree++;
		   }
	   }
	   
	   public IDiagram getIDiagram(){
		   return this.iDiagram;
	   }
	   public ILoopIndex getILoopIndex(){
		   return this.iLoopIndex;
	   }
	   public ILoopMax getILoopMax(){
		   return this.iLoopMax;
	   }
	   public Vector<ITunnel> getITunnels(){
		   return this.iTunnels;
	   }
	   public Vector<ILeftShiftRegister> getILeftShiftRegisters(){
		   return this.iLeftShiftRegisters;
	   }
	   public Vector<IRightShiftRegister> getIRightShiftRegisters(){
		   return this.iRightShiftRegisters;
	   }
	   public void print(){
		   System.out.println("IForLoop begin");
		   System.out.println("NodeId: " + this.nodeId + ' ' + "ParentId: " + this.parentId);
		   this.iLoopIndex.print();
		   this.iLoopMax.print();
		   for(int i=0; i<this.iTunnels.size(); i++){
			   this.iTunnels.get(i).print();
		   }
		   for(int i=0; i<this.iLeftShiftRegisters.size(); i++){
			   this.iLeftShiftRegisters.get(i).print();
		   }
		   for(int i=0; i<this.iRightShiftRegisters.size(); i++){
			   this.iRightShiftRegisters.get(i).print();
		   }
		   this.iDiagram.print();
		   System.out.println("IForLoop end\n");
	   }
	   public Vector<JNode> getOutputBorderNodes(){
		      return this.outputBorderNodes;
	   }
	   public Vector<JNode> getInputBorderNodes(){
			  return this.inputBorderNodes;
	   }
}
